### PR TITLE
Improve exceptions emitted and handled

### PR DESCRIPTION
**Pull request by Jacob-Stevens-Haas**
No more naked exceptions, except in one place where it truly is intended (i.e. if something bad happens, warn about it and continue the loop).  Exception testing rewritten in pytest, since that's where we're headed and I don't know how to modify robot tests.

* Changes _printErrorMessage to _createErrorMessage.  Exception handling
determines whether to print error message
* moved MaxRetriesException to _DataProductFile.py, its only user
* saveAsFile no longer handles error codes
* added some type annotations
* modify ONC to accept Path objects for outPath
* Modify tests/readme.md for pytest tests

Worth noting that documentation on wiki is unclear whether some error codes from server should be 127/129.
(e.g. test_raw_bad_filters/test_bad_filters).  Other services do not list error codes on the ONC wiki.

### CODE REVIEW

**Comment by kan-fu** [comment]
Monday and Tuesday are our regression test days before the minor release. We will come back to the PR on Wednesday.

**Comment by Jacob-Stevens-Haas** [comment]
> ...We will come back to the PR on Wednesday.
> 
> BTW I think it would be better to merge your PR first....

Sorry, I'm confused.  Were you approving this PR and saying that we'll come back to your PR on Wednesday, or were you saying that we'll come back to this PR on Wednesday and then your PR?

File: src/onc/modules/_DataProductFile.py [hunk 1]
```diff
@@ -48,80 +53,52 @@ def download(
...
+            elif self._status == 400:
+                raise requests.HTTPError(_createErrorMessage(response))
+
+            elif self._status == 404:  # Index too high, no more files to download
+                log.printNewLine()
+                pass
+
+            elif self._status == 410:  # Status 410: gone (file deleted from FTP)
+                warn(
+                    "   FTP Error: File not found. If the product order is recent,",
```
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply src/onc/modules/_DataProductFile.py:98]: You didn't leave any confusion, I just made a mistake :)

File: src/onc/onc.py [hunk 2]
```diff
@@ -32,6 +33,7 @@ def __init__(
         self.baseUrl = "https://data.oceannetworks.ca/"
         self.outPath = ""
 
+        outPath = str(outPath)
```
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply src/onc/onc.py:36]: That's a good idea.

File: src/onc/modules/_util.py [hunk 3]
```diff
@@ -66,18 +61,16 @@ def _printErrorMessage(response):
                 code = e["errorCode"]
                 msg = e["errorMessage"]
                 parameters = e["parameter"]
-                print(f"   Error {code}: {msg} (parameter: {parameters})")
+                return f"   Error {code}: {msg} (parameter: {parameters})"
```
[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply src/onc/modules/_util.py]: All great points.  
1. Incorporated line 57, delayed the return until the end of the loop, and accumulated errors
2. Yep, if the API returns a list here, we don't need to guard this in a conditional.
3. Agree, only difference is that I called the codes that the API generates `API Error {code}`.  This way, the error appears as:
```
HTTPError: Status 400 - Bad Request: oceannetworks.ca/...<api endpoint>
API Error: 127 ( ...
API Error: 129 (...
```
Made the change for HTTP code 401, too.

**Comment by kan-fu** [comment]
LGTM! Thanks for your effort. This saves me a lot of trouble when reading `except Exception: raise` pattern.

One small thing, if you find splitting a long string makes the string look awkward, just add `# noqa: E501` to disable the linting. I added several myself, and tend not to be too strict about this [line-too-long](https://docs.astral.sh/ruff/rules/line-too-long/) rule. 

For the linting, don't worry about it. We will merge yours first. As I said earlier, something is not set up correctly between isort and ruff check I rule. I will fix it in #19.

**Review by kan-fu — APPROVED** [review]

**Comment by Jacob-Stevens-Haas** [comment]
Great, thanks Kan!  I tend to be pretty strict about splitting strings unless it's something like a URL that will be copy/pasted a lot.

Am I able to merge this, or are we waiting on @spencerwplovie and @eliferguson?

**Review by eliferguson — APPROVED** [review]

**Pull request merged**